File: backend/app/schemas/claim.py
"""Claim flow (/claim) — see docs/API_CONTRACTS.md "Claim flow (/claim)"."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, model_validator

ProofMethod = Literal["google_business_profile", "phone_verification", "document_upload"]


class ClaimCreate(BaseModel):
    brand_id: int
    location_id: int | None = None
    proof_method: ProofMethod
    google_business_profile_url: str | None = None
    supporting_document_url: str | None = None

    @model_validator(mode="after")
    def _validate_proof(self) -> "ClaimCreate":
        if self.proof_method == "google_business_profile" and not self.google_business_profile_url:
            raise ValueError("google_business_profile_url is required for this proof method")
        if self.proof_method == "document_upload" and not self.supporting_document_url:
            raise ValueError("supporting_document_url is required for this proof method")
        return self


class ClaimOut(BaseModel):
    claim_id: int
    brand_id: int
    status: str
    proof_method: str
    submitted_at: datetime
    sla_due_at: datetime
    reviewed_at: datetime | None = None
    reviewer_notes: str | None = None
    # Only ever set on the `POST /claim/{id}/approve` response: True = claimant
    # added to the Cognito `owner` group, False = attempted and failed
    # (approval itself still committed), None = not applicable/not attempted.
    owner_group_granted: bool | None = None


ClaimStatus = Literal["pending_review", "approved", "rejected"]


class ClaimQueueItem(BaseModel):
    """Admin-facing claim row (`GET /claim`). `supporting_document_url` is
    the stored S3 key (no presigned read URL is minted here)."""

    claim_id: int
    brand_id: int
    brand_name: str
    brand_slug: str
    location_id: int | None = None
    location_address: str | None = None
    claimant_user_id: str
    claimant_email: str | None = None
    proof_method: str
    google_business_profile_url: str | None = None
    supporting_document_url: str | None = None
    status: str
    submitted_at: datetime
    sla_due_at: datetime
    reviewed_at: datetime | None = None
    reviewer_notes: str | None = None


class ClaimListResponse(BaseModel):
    results: list[ClaimQueueItem]
    page: int
    page_size: int
    total: int


class ClaimApproveRequest(BaseModel):
    reviewer_notes: str | None = None


class ClaimRejectRequest(BaseModel):
    reviewer_notes: str

File: frontend/src/types/claim.ts
// Types for POST /claim and its follow-ups, matching docs/API_CONTRACTS.md
// "Claim flow (`/claim`)" and docs/DECISIONS.md "Claim flow".

export type ClaimProofMethod =
  | "google_business_profile"
  | "phone_verification"
  | "document_upload";

export type ClaimStatus = "pending_review" | "approved" | "rejected";

/** Body for POST /claim. */
export interface CreateClaimInput {
  brand_id: number;
  /**
   * Required in practice for proof_method="phone_verification" when the
   * brand has more than one location (identifies which location's public
   * phone number is being called). Omit for a single-location brand or the
   * other two proof methods.
   */
  location_id?: number;
  proof_method: ClaimProofMethod;
  google_business_profile_url?: string | null;
  supporting_document_url?: string | null;
}

/** Response for POST /claim and GET /claim/{id}. */
export interface ClaimResponse {
  claim_id: number;
  brand_id: number;
  status: ClaimStatus;
  proof_method: ClaimProofMethod;
  submitted_at: string;
  /** submitted_at + 2 business days, computed, not stored. */
  sla_due_at: string;
  reviewed_at?: string | null;
  reviewer_notes?: string | null;
}

/**
 * Row of GET /claim (admin claims queue). `supporting_document_url` is the
 * stored S3 key as-is — the API mints no presigned read URL.
 */
export interface ClaimQueueItem {
  claim_id: number;
  brand_id: number;
  brand_name: string;
  brand_slug: string;
  location_id: number | null;
  location_address: string | null;
  claimant_user_id: string;
  /** null when no owner_account row matches the claimant. */
  claimant_email: string | null;
  proof_method: ClaimProofMethod;
  google_business_profile_url: string | null;
  supporting_document_url: string | null;
  status: ClaimStatus;
  submitted_at: string;
  sla_due_at: string;
  reviewed_at: string | null;
  reviewer_notes: string | null;
}

/** Body for POST /claim/{id}/approve. Every field optional. */
export interface ApproveClaimInput {
  reviewer_notes?: string;
}

/** Body for POST /claim/{id}/reject. reviewer_notes is required to reject. */
export interface RejectClaimInput {
  reviewer_notes: string;
}
